### PR TITLE
Improve 2D editor zoom logic

### LINKED ISSUE #2658: Improve 2D editor zoom controls
### Describe the project you are working on
A 2D shootemup game.

### Describe the problem or limitation you are having in your project
I am unable to set the Godot editor's zoom level how I want, quickly and easily.
Sometimes I want to set Godot's zoom so it matches my game's scale, while I'm making levels.
This helps a ton in giving me a good sense of how things look and feel on-screen when playing the game (without needing to run the game of course).
It's a very basic, natural desire for a level designer to want to have a good sense of what his level looks like, as he's making it (as much as possible).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Because I know my game is rendered at a certain resolution and is blown up to fill my screen, if I zoom to 500% while working on a new level, it perfectly matches the run-time zoom level when playing the game. Godot's zoom function doesn't facilitate this at all.

It could, though. A small change could mean a nice increase in user-friendliness.

Solutions? I have three ideas for you guys:
**IDEA 01)** Add new zoom shortcuts so users can jump to 200%, 300%, 400%, etc, in one keystroke.
Add shortcut slots for 200%, 300%, 400%, 500%, 600%, 700%, and 800% to Editor Settings.

I recommend _not_ setting default keystrokes for new these keyboard shortcut slots, since users may already be using the keystrokes. 

We only have these currently:
(https://i.imgur.com/Avhxh0L.png)
Screenshot of current Godot zoom keyboard shortcut slots.

Something like this would be absolutely ideal:
(https://i.imgur.com/orf9w6L.png)
Screenshot of zoom keyboard shortcuts.

If I could simply press CTRL+5 to go to right to 500% zoom, it would be great!

**IDEA 02)** Standardize Godot's zoom level snapping when using scrollwheel or clicking plus/minus buttons.
Lock zoom increments to be more predictable whole number increments, or let users hold a modifier key when zooming that would do this.   ie  ALT+SCROLLWHEEL rounds to nearest 100% increment then each ALT_SCROLLWHEEL tick thereafter snaps to 100's.
As user continues to zoom out beyond 100%, then decrease the zoom snap increments.
_For reference, Photoshop snaps to 100% increments just like I'm describing, and once zoomed to 100% and user continues to zoom out then zoom snaps to these numbers in this order: 66%, 50%, 33%, 25%, 16.5%, 12.5%, 8.3%, 6.25%, 5%, and so on._

Currently, Godot 3.3's zooming is unpredictable and seems a little random:
[-Screen recording of Godot 3.3 while zooming-](https://i.imgur.com/z25XcnB.gif)
I can't get to 500% even if I want to! This does not seem right at all. It's a minor matter but honestly it feels dysfunctional, to me.

**IDEA 03)** Let users right-click on zoom percentage number (top left of editor) and input a zoom percentage and press ENTER to jump to that exact zoom level. (currently RMB on the top left zoom percent does nothing)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
_(N/A? Please see above; I don't want to be overly verbose and waste your time)_

TL;DR: With this idea in place, users would be able to use a single keystroke to jump straight to very common zoom levels, or use their mouse scrollwheel to do so, if holding a modifier key.
At the least - add new shortcut slots users can set that affect the editor's zoom level.
At the most - change the math used for calculating zoom level changes so that simpler whole number increments are used.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
No. It concerns modifying the editor, fundamentally.

### Is there a reason why this should be core and not an add-on in the asset library?
Yes. This proposal is suggesting unobtrusive, optional enhancements for the Godot Editor which can be ignored.
In my opinion, this change would be embraced as purely beneficial by a large percentage of 2D users.

### PR DESCRIPTION
**Pull request by Calinou**
`master` version of https://github.com/godotengine/godot/pull/50490.

- Add <kbd>1</kbd>-<kbd>5</kbd> shortcuts to zoom between 100% and 1600% quickly (similar to GIMP).
- When holding down <kbd>Alt</kbd>, go through integer zoom values if above 100% or fractional zoom values with integer denominators if below 100% (50%, ~33.3%, 25%, …).

This closes https://github.com/godotengine/godot-proposals/issues/2658.

### CODE REVIEW

**Comment by mathias17** [comment]
OK this is even better - simply pressing 1-9, without using Control, to alter zoom.

Two questions:
1. Will these 9 new shortcuts be added to the Shortcuts tab in Editor Settings?
2. Will these 9 new shortcuts work no matter which panel currently has focus?

**Comment by Calinou** [comment]
> 1. Will these 9 new shortcuts be added to the Shortcuts tab in Editor Settings?

Yes, this is already done.

> 2. Will these 9 new shortcuts work no matter which panel currently has focus?

No, these only work if the 2D editor viewport is focused.

**Comment by groud** [comment]
I am not really sold on the proposal. This is a lot of shortcuts for something that simply requires pressing Ctrl+'+' or Ctrl+'-' otherwise. And I'm not sure we need that amount of different values with one key each.

I think it would make more sense to make those 1-9 keys switch between scenes for example.

**Comment by mathias17** [comment]
> I am not really sold on the proposal. This is a lot of shortcuts for something that simply requires pressing Ctrl+'+' or Ctrl+'-' otherwise. And I'm not sure we need that amount of different values with one key each.
> 
> I think it would make more sense to make those 1-9 keys switch between scenes for example.

Interesting idea for number row keys 1 through 0 to be mapped to scene tabs. I think I would actually use that.
This type of shortcut functionality doesn't seem like the user would be able to modify it. I guess it would just be an arbitrary feature that users wouldn't be able to modify/turn off - which is fine.

If the consensus is for 1 - 0 to be mapped to scene tabs, the I still suggest CTRL+[number] for specific zoom shortcuts.

And at the very least, as part of this proposal - ALT+SCROLLWHEEL will snap to integer zoom values. A big improvement, imo.

**Comment by KoBeWi** [comment]
> I think it would make more sense to make those 1-9 keys switch between scenes for example.

The difference is that there's no way to jump to zoom values, while you have all scene tabs available at once. No reason to switch them with number if you can just click them.

That said, the "1" shortcut is the same as clicking zoom label, but personally I also use it to reset zoom 🤔

**Comment by Calinou** [comment]
> I know noticed that there is both "Reset Zoom" and "Zoom 100%". I think the former should be removed.

We need both shortcuts until support for multiple editor shortcuts is implemented.

**Comment by KoBeWi** [comment]
I mean, if Reset Zoom and Zoom 100% is the same thing, we don't need both.

**Comment by Calinou** [comment]
> I'd just remove Ctrl+0, as 1 is easier to use anyways.

Most applications such as web browsers and code editors support Ctrl + 0 to reset the zoom.

**Comment by KoBeWi** [comment]
Wow, that Alt-zoom is so nice to use.

I still think that having two functionally identical shortcuts is redundant. We could remove it for now and possibly re-add when the multi-event shortcuts are added.

Also, this is inspired by GIMP, but GIMP has actually 5 zoom shortcuts. 1 for 100%, 2 for 200%, 3 for 400%, 4 for 800 and 5 for 1600%. I tried the numerical zooms and, while this is handy, having 9 of them, with 100% increment isn't very useful (especially when Alt+wheel does exactly that). Maybe it could be changed to something closer to GIMP, i.e. less shortcuts and more exponential scale?

**Comment by Calinou** [comment]
> Also, this is inspired by GIMP, but GIMP has actually 5 zoom shortcuts. 1 for 100%, 2 for 200%, 3 for 400%, 4 for 800 and 5 for 1600%. I tried the numerical zooms and, while this is handy, having 9 of them, with 100% increment isn't very useful (especially when Alt+wheel does exactly that). Maybe it could be changed to something closer to GIMP, i.e. less shortcuts and more exponential scale?

I thought about doing this at first, but I feel like zoom values above 800% aren't often useful in game development. (This is different when editing pixel art, for instance.)

**Comment by KoBeWi** [comment]
Then it would stop at `4` / 800%.

**Comment by Calinou** [comment]
> Then it could stop at `4` / 800%.

Well, adding `5` / 1600% isn't a problem, it's just that being able to jump to any 100% increment quickly feels better overall.

**Comment by KoBeWi** [comment]
But Alt + wheel already gives 100% increment. Having 9 shortcuts for that would be only useful if you use keyboard only, which is unlikely in the editor. I use GIMP quite a bit and IMO its zoom shortcut scheme is enough.

> Adding 5 / 1600% isn't a problem

I know. That's why we could just have 4 shortcuts. Or also 5, but with different increments, e.g. 100/200/400/600/800.

**Comment by Calinou** [comment]
I changed the zoom shortcuts to use power-of-two integer factors between 100% and 1600%.

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 1]
```diff
@@ -5610,6 +5626,16 @@ CanvasItemEditor::CanvasItemEditor(EditorNode *p_editor) {
...
 
+	// To ensure that scripts can parse the list of shortcuts correctly, we have to define
+	// those shortcuts one by one.
+	// Resetting zoom to 100% is a duplicate shortcut of `canvas_item_editor/reset_zoom`,
+	// but it ensures both 1 and Ctrl + 0 can be used to reset zoom.
+	ED_SHORTCUT("canvas_item_editor/zoom_100_percent", TTR("Zoom To 100%"), KEY_1);
+	ED_SHORTCUT("canvas_item_editor/zoom_200_percent", TTR("Zoom To 200%"), KEY_2);
+	ED_SHORTCUT("canvas_item_editor/zoom_400_percent", TTR("Zoom To 400%"), KEY_3);
+	ED_SHORTCUT("canvas_item_editor/zoom_800_percent", TTR("Zoom To 800%"), KEY_4);
+	ED_SHORTCUT("canvas_item_editor/zoom_1600_percent", TTR("Zoom To 1600%"), KEY_5);
```
Comment (akien-mga) [inline editor/plugins/canvas_item_editor_plugin.cpp:5637]: I think the correct capitalization would be "Zoom to 100%", etc.

https://titlecaseconverter.com/blog/is-to-capitalized/

Comment (Calinou) [reply editor/plugins/canvas_item_editor_plugin.cpp:5637]: Done: https://github.com/godotengine/godot/pull/50418

**Comment by akien-mga** [comment]
Could probably use a dedicated PR for `3.x`, it's not trivial to cherry-pick.
